src/blocking.rs

```diff
@@ -0,0 +1,2830 @@
+pub fn current_platform_resolver(manifests: &[ImageIndexEntry]) -> Option<String> {
+    manifests
+        .iter()
+        .find(|entry| {
+            entry.platform.as_ref().map_or(false, |platform| {
+                platform.os == go_os() && platform.architecture == go_arch()
+            })
+        })
+        .map(|entry| entry.digest.clone())
+}
+
+/// The protocol that the client should use to connect
+#[derive(Debug, Clone, PartialEq, Eq, Default)]
+pub enum ClientProtocol {
+    #[allow(missing_docs)]
+    Http,
+    #[allow(missing_docs)]
+    #[default]
+    Https,
+    #[allow(missing_docs)]
+    HttpsExcept(Vec<String>),
+}
+
+impl ClientProtocol {
+    fn scheme_for(&self, registry: &str) -> &str {
+        match self {
+            ClientProtocol::Https => "https",
+            ClientProtocol::Http => "http",
+            ClientProtocol::HttpsExcept(exceptions) => {
+                if exceptions.contains(&registry.to_owned()) {
+                    "http"
+                } else {
+                    "https"
+                }
+            }
+        }
+    }
+}
+
+#[derive(Clone, Debug)]
+struct BearerChallenge {
+    pub realm: Box<str>,
+    pub service: Option<String>,
+}
+
+impl TryFrom<&HeaderValue> for BearerChallenge {
+    type Error = String;
+
+    fn try_from(value: &HeaderValue) -> std::result::Result<Self, Self::Error> {
+        let parser = ChallengeParser::new(
+            value
+                .to_str()
+                .map_err(|e| format!("cannot convert header value to string: {:?}", e))?,
+        );
+        parser
+            .filter_map(|parser_res| {
+                if let Ok(chalenge_ref) = parser_res {
+                    let bearer_challenge = BearerChallenge::try_from(&chalenge_ref);
+                    bearer_challenge.ok()
+                } else {
+                    None
+                }
+            })
+            .next()
+            .ok_or_else(|| "Cannot find Bearer challenge".to_string())
+    }
+}
+
+impl TryFrom<&ChallengeRef<'_>> for BearerChallenge {
+    type Error = String;
+
+    fn try_from(value: &ChallengeRef<'_>) -> std::result::Result<Self, Self::Error> {
+        if !value.scheme.eq_ignore_ascii_case("Bearer") {
+            return Err(format!(
+                "BearerChallenge doesn't support challenge scheme {:?}",
+                value.scheme
+            ));
+        }
+        let mut realm = None;
+        let mut service = None;
+        for (k, v) in &value.params {
+            if k.eq_ignore_ascii_case("realm") {
+                realm = Some(v.to_unescaped());
+            }
+
+            if k.eq_ignore_ascii_case("service") {
+                service = Some(v.to_unescaped());
+            }
+        }
+
+        let realm = realm.ok_or("missing required parameter realm")?;
+
+        Ok(BearerChallenge {
+            realm: realm.into_boxed_str(),
+            service,
+        })
+    }
+}
+
+#[cfg(test)]
+mod test {
```

Comment:
I would really prefer if we could find a way to test both clients using the same tests. Even if we just had an enum with the blocking and non blocking client (that we only used for tests) that matches on the client and calls the right thing